src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp

```diff
@@ -505,159 +505,228 @@ real Hydro_CheckMinEintInEngy( const real Dens, const real MomX, const real MomY
+            else
+            {
+               if ( (real) Fields[v] <=  TINY_NUMBER || Fields[v]  >= (real)HUGE_NUMBER )
+                  FailCell = true;
             }
-#           endif
          }
-         else if ( Cons != NULL && Prim == NULL && Input == NULL )
+
+
+         // calculate discriminant for SRHD.
+         // --> the discriminant is positive if and only if the Eq.(15) in MNRAS 504, 3298–3315 (2021) has a positive root.
+#        ifdef SRHD
+         Msqr         = SQR(Fields[MOMX]) + SQR(Fields[MOMY]) + SQR(Fields[MOMZ]);
+         Dsqr         = SQR(Fields[DENS]);
+         E_D          = Fields[ENGY] / Fields[DENS];
+         M_D          = SQRT( Msqr / Dsqr );
+         Temp         = SQRT( E_D*E_D + (real)2.0*E_D );
+         Discriminant = ( Temp + M_D ) * ( Temp - M_D );
+
+         // check discriminant
+         if ( Discriminant <= TINY_NUMBER )      FailCell = true;
+#        endif
+
+
+         // print out the unphysical values
+         if ( FailCell && Verbose )
          {
             printf( "ERROR: unphysical conserved variables at file <%s>, line <%d>, function <%s>\n",
-                            File, Line, Function );
+                     File, Line, Function );
+
+            printf( "D=%14.7e, Mx=%14.7e, My=%14.7e, Mz=%14.7e, E=%14.7e\n",
+                     Fields[DENS], Fields[MOMX], Fields[MOMY], Fields[MOMZ], Fields[ENGY] );
+
 #           ifdef SRHD
-            printf( "       D=%14.7e, Mx=%14.7e, My=%14.7e, Mz=%14.7e, E=%14.7e, E^2+2*E*D-|M|^2=%14.7e\n",
-                            Cons[DENS], Cons[MOMX], Cons[MOMY], Cons[MOMZ], Cons[ENGY], Discriminant );
-#           else
-            printf( "       D=%14.7e, Mx=%14.7e, My=%14.7e, Mz=%14.7e, E=%14.7e\n",
-                            Cons[DENS], Cons[MOMX], Cons[MOMY], Cons[MOMZ], Cons[ENGY] );
+            printf( "E^2+2*E*D-|M|^2=%14.7e\n", Discriminant );
+#           endif
+
+#           if ( NCOMP_PASSIVE > 0 )
+            for (int v=NCOMP_FLUID; v<NCOMP_TOTAL; v++)
+               printf("Passive[%d]=%14.7e\n", v, Fields[v] );
 #           endif
          }
-         else if ( Cons == NULL && Prim != NULL && Input == NULL )
+
+         return FailCell;
+
+      break;
+
+
+      // === check primitive variables, including passive scalars  ===
+      case UNPHY_MODE_PRIM:
+
+         for (int v=0; v<NCOMP_TOTAL; v++)
+         {
+            // check NaN
+            if ( Fields[v] != Fields[v] )
+                  FailCell = true;
+
+            // check momentum densities
+            if ( v == MOMX || v == MOMY || v == MOMZ )
+            {
+               if ( (real) Fields[v] <= -HUGE_NUMBER || Fields[v]  >= (real)HUGE_NUMBER )
+                  FailCell = true;
+            }
+            // otherwise, including passive scalars
+            else
+            {
+               if ( (real) Fields[v] <=  TINY_NUMBER || Fields[v]  >= (real)HUGE_NUMBER )
+                  FailCell = true;
+            }
```

Comment:
Strictly speaking, passive scalars can be zero. So replace it with the following.
```
// check mass and pressure
else if ( v < NCOMP_TOTAL ) {
   if ( Fields[v] <= TINY_NUMBER  ||  Fields[v] >= HUGE_NUMBER )  FailCell = true;
}  
            
// check passive scalars
else {
   if ( Fields[v] < (real)0.0  ||  Fields[v] >= HUGE_NUMBER )     FailCell = true;
}
```